cou/apps/auxiliary.py

```diff
@@ -0,0 +1,165 @@
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: The next channel for the application. E.g: victoria/stable
+        :rtype: str
+        """
+        return f"{self.openstack_to_track(target)}/stable"
+
+    def _get_refresh_charm_plan(
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan for refreshing the current channel.
+
+        This function also identifies if charm comes from charmstore and in that case,
+        makes the migration.
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan for refreshing the charm.
+        :rtype: Optional[UpgradeStep]
+        """
+        switch = None
+        channel = self.channel
+
+        description = f"Refresh '{self.name}' to the latest revision of '{self.channel}'"
+
+        if self.charm_origin == "cs":
+            description = f"Migration of '{self.name}' from charmstore to charmhub"
+            switch = f"ch:{self.charm}"
+            channel = self.expected_current_channel
+        elif self.channel != self.expected_current_channel:
+            logger.warning(
+                "'%s' has the channel set to: %s which is different from the expected channel: %s",
+                self.name,
+                self.channel,
+                self.expected_current_channel,
+            )
+            return None
+
+        # pylint: disable=duplicate-code
+        return UpgradeStep(
+            description=description,
+            parallel=parallel,
+            function=async_upgrade_charm,
+            application_name=self.name,
+            channel=channel,
+            model_name=self.model_name,
+            switch=switch,
+        )
+
+    def upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Upgrade planning.
+
+        Auxiliary charms have multiple compatible OpenStack releases. In that case,
+        we check also the OpenStack origin from the "source" or "openstack-origin" to know
+        if it's necessary to change it. E.g: rabbitmq-server with workload version 3.8 is
+        compatible from ussuri to yoga. Even that is considered as yoga, we need to set the
+        source accordingly with the OpenStack components that might be in a version lower than
+        yoga.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation
+        :return: Plan that will add upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        os_origin_config = self.os_origin_config(target)
+        if self.current_os_release >= target and os_origin_config >= target:
```

Comment:
why was it not necessary to check the origin setting for OpenStack control plane apps?

Reply:
I tried to explain in the docstring, but basically on the control plane apps we have a 1:1 map between workload_version and OpenStack release. On the auxiliary charms we don't. E.g: version 3.8 on rabbitmq-server compatible with ussuri to yoga. In that case, if we want to set `source` of rmq to the apt repository of the current cloud version, we need to know where the "source" is pointing.

Moreover, we are checking if the workload has been upgraded. So in case a control plane changes the `openstack-origin` and the workload doesn't change, an error will happen.

Reply:
However, this logic on OpenStack control plane would make no harm and we can use the same method for both classes. It's actually a good idea passing it to the super class and not need to overwrite the method